### PR TITLE
feat: Add support for cross-platform Data Collection Rules

### LINKED ISSUE #3071: [AVM Module Issue]: DCR 'kind' should allow being undefined to support both Windows and Linux.
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Bug

### Module Name

avm/res/insights/data-collection-rule

### (Optional) Module Version

0.2.0

### Description

Property `Kind` is optional but defined with default value `Linux` in the template. 
This property should be optional without a default value to support xPlat (all) scenarios. 

image

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by krbar**
## Description

Adding support for cross-platform Data Collection Rules (Windows + Linux). 

Use the parameter `dataCollectionRuleProperties.kind: 'All'` 

Fixes #3071 

## Pipeline Reference

| Pipeline |
| -------- |
|    [avm.res.insights.data-collection-rule](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.insights.data-collection-rule.yml)      |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [x] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [x] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [x] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/res/insights/data-collection-rule/main.bicep [hunk 1]
```diff
@@ -75,78 +70,61 @@ resource avmTelemetry 'Microsoft.Resources/deployments@2024-03-01' = if (enableT
...
+// Using a separate resource for parameter kind: 'All' as it requires that the "kind' is not set on the resource and 'kind: null' is not allowed for this resource type
+resource dataCollectionRuleAll 'Microsoft.Insights/dataCollectionRules@2023-03-11' = if (dataCollectionRuleProperties.kind == 'All') {
+  location: location
+  name: name
+  tags: tags
+  properties: dataCollectionRulePropertiesUnion
+}
+
+// Using a module as a workaround for issues with conditional scope: https://github.com/Azure/bicep/issues/7367
+module dataCollectionRule_lock 'modules/nested_lock.bicep' = if (!empty(lock ?? {}) && lock.?kind != 'None') {
```
Comment (AlexanderSehr) [inline avm/res/insights/data-collection-rule/main.bicep:90]: I'm sure this works - but - I can't help but wonder if we should not have both the lock & roleAssignment in one nested file. This not only simplifies this template, but also means 1 less deployment.

File: avm/res/insights/data-collection-rule/modules/nested_lock.bicep [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      ? 'Cannot delete resource or child resources.'
+      : 'Cannot delete or modify the resource or child resources.'
+  }
+}
+
+// =============== //
+//   Definitions   //
+// =============== //
+
+type lockType = {
```
Comment (AlexanderSehr) [inline avm/res/insights/data-collection-rule/modules/nested_lock.bicep:26]: If youi define the type here too, I'd advice to add an `export()` flag and import it in the parent. Then you have one less place to maintain it from.

File: avm/res/insights/data-collection-rule/modules/nested_roleAssignments.bicep [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+param roleAssignments roleAssignmentType
+
+@description('Optional. Name of the Data Collection Rule to assign the role(s) to.')
+param dataCollectionRuleName string
+
+var builtInRoleNames = {
+  Contributor: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'b24988ac-6180-42a0-ab88-20f7382dd24c')
+  Owner: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8e3af657-a8ff-443c-a75c-2fe8c4bcb635')
+  Reader: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
+  'Role Based Access Control Administrator (Preview)': subscriptionResourceId(
```
Comment (AlexanderSehr) [inline avm/res/insights/data-collection-rule/modules/nested_roleAssignments.bicep:11]:
```suggestion
  'Role Based Access Control Administrator': subscriptionResourceId(
```
While we're at it ...

Comment (AlexanderSehr) [reply avm/res/insights/data-collection-rule/modules/nested_roleAssignments.bicep:11]: On a different note, I wonder if we should keep the roles and formatting in the parent. The only reason I say so is that users need to know they exist to actually set them, which gets harder if they're in a nested file. Also, assuming I ever get somebody to help me merge [this](https://github.com/Azure/bicep-registry-modules/pull/706) PR, the roles would need to be in the parent... I think

File: avm/res/insights/data-collection-rule/modules/nested_lock.bicep (hunk 2 above)
Comment (krbar) [reply avm/res/insights/data-collection-rule/modules/nested_lock.bicep]: addressed in: https://github.com/Azure/bicep-registry-modules/pull/3211/commits/dd9ddf789ccb41bf73fa82992e84dfeb59ed51bd

File: avm/res/insights/data-collection-rule/modules/nested_roleAssignments.bicep (hunk 3 above)
Comment (krbar) [reply avm/res/insights/data-collection-rule/modules/nested_roleAssignments.bicep]: Addressed in: https://github.com/Azure/bicep-registry-modules/pull/3211/commits/dd9ddf789ccb41bf73fa82992e84dfeb59ed51bd

**Comment by krbar** [comment]
@AlexanderSehr thank you for the quick review! The comments are addressed and the module re-tested.

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**